### PR TITLE
Sttp4 generator

### PR DESCRIPTION
**Pull request by flsh86**
An implementation of generator for scala sttp4 client (ref: https://sttp.softwaremill.com/en/latest/index.html)

```
git clone https://github.com/OpenAPITools/openapi-generator/
cd openapi-generator
git checkout sttp4-generator

Linux: ./mvnw clean package
Windows: .\mvnw.bat clean package

java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -i https://raw.githubusercontent.com/openapitools/openapi-generator/master/modules/openapi-generator/src/test/resources/3_0/petstore.yaml -g scala-sttp4

```

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] In case you are adding a new generator, run the following additional script : 
  ```
  ./bin/utils/ensure-up-to-date
  ``` 
  Commit all changed files.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.3.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by wing328** [comment]
/
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/OpenAPITools/openapi-generator/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

**Comment by wing328** [comment]
cc @clasnake (2017/07), @jimschubert (2017/09) ❤️, @shijinkui (2018/01), @ramzimaalej (2018/03), @chameleon82 (2020/03), @Bouillie (2020/04)

File: modules/openapi-generator/src/main/resources/scala-sttp4/api.mustache [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{{>licenseInfo}}
+package {{package}}
+
+{{#imports}}
+    import {{import}}
```
Comment (chameleon82) [inline modules/openapi-generator/src/main/resources/scala-sttp4/api.mustache:5]: formatting. unnecessary blank spaces

File: modules/openapi-generator/src/main/resources/scala-sttp4/api.mustache [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+{{/imports}}
+import {{invokerPackage}}.JsonSupport._
+import sttp.client4._
+import sttp.model.Method
+
+{{#operations}}
+    object {{classname}} {
```
Comment (chameleon82) [inline modules/openapi-generator/src/main/resources/scala-sttp4/api.mustache:12]: formatting. redundant spaces

File: modules/openapi-generator/src/main/resources/scala-sttp4/api.mustache [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    def apply(baseUrl: String = "{{{basePath}}}") = new {{classname}}(baseUrl)
+    }
+
+    class {{classname}}(baseUrl: String) {
+
+    {{#operation}}
+        {{#javadocRenderer}}
+            {{>javadoc}}
+        {{/javadocRenderer}}
+        def {{operationId}}({{>methodParameters}}): Request[{{#defaultReturnType}}Either[Either[String, String], Unit]{{/defaultReturnType}}{{^defaultReturnType}}{{#separateErrorChannel}}Either[ResponseException[String, Exception], {{>operationReturnType}}]{{/separateErrorChannel}}{{^separateErrorChannel}}{{>operationReturnType}}{{/separateErrorChannel}}{{/defaultReturnType}}] =
```
Comment (chameleon82) [inline modules/openapi-generator/src/main/resources/scala-sttp4/api.mustache:23]: formatting. here and bellow. please fix with default formatting rules

File: samples/client/petstore/scala/sttp4/project/build.properties [hunk 4]
```diff
@@ -0,0 +1 @@
+sbt.version=1.6.1
```
Comment (chameleon82) [inline samples/client/petstore/scala/sttp4/project/build.properties:1]: might be updated to 1.9.0 ?

**Review by chameleon82 — COMMENTED** [review]
Please fix formatting, the rest looks good to me

**Comment by flsh86** [comment]
> Please fix formatting, the rest looks good to me

I've changed files accordingly to accepted formatting. Please take a look if it's good now

Thanks

File: samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+  /**
+   * 
+   * 
+   * Expected answers:
+   *   code 400 :  (Invalid pet value)
+   * 
+   * @param petId Pet id to delete
+   * @param apiKey 
+   */
+  def deletePet(petId: Long, apiKey: Option[String] = None
```
Comment (chameleon82) [inline samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala:54]: Formatting issue

File: samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+import org.openapitools.client.model.ApiResponse
+import java.io.File
+import org.openapitools.client.model.Pet
+import org.openapitools.client.core.JsonSupport._
+import sttp.client4._
+import sttp.model.Method
+
+object PetApi {
+
+def apply(baseUrl: String = "http://petstore.swagger.io/v2") = new PetApi(baseUrl)
```
Comment (chameleon82) [inline samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala:23]: minor. formatting issue.

**Review by chameleon82 — APPROVED** [review]

**Comment by wing328** [comment]
@flsh86 thanks again for the PR. Have you considered updating the existing `scala-sttp` client generator to sttp 4.x instead of creating a new generator?

or the new generator is actually based on `scala-sttp` (java class and templates) ?

**Comment by chameleon82** [comment]
sttp4 has different package, but @wing328 point is valid. sttp4 is just major update of the same client (they have some incompatible changes, but api remain the same).
Imo sttp4 still M1, but if they release stable version it can be just a major update released in `7.0.0` as `sttp` instead of separate `sttp4` template

**Comment by adamw** [comment]
@chameleon82 yes, as you right this is a major update, however at least for some time I suspect people will be using both 3.x and 4.x versions. They can be used side-by-side though, as the packages are different, so in a way both of these projects are independent.

**Comment by wing328** [comment]
> at least for some time I suspect people will be using both 3.x and 4.x versions

Very good point.

What about cloning the existing scala-sttp to create scala-sttp4 so that all previous enhancements and fixes can carry over to scala-sttp4?

If scala-sttp4 is built from scratch, we may encounter some issues already addressed in scala-sttp.

**Comment by adamw** [comment]
@wing328 I think that's exactly what @flsh86 did :)

**Comment by wing328** [comment]
@adamw nice 👍

**Comment by wing328** [comment]
tested locally and the result is good
```
  100.0% [##########] 5.7 MiB (725.9 KiB / s)
https://repo1.maven.org/maven2/org/scala-lang/scala-compiler/2.13.10/scala-compiler-2.13.10.jar
  100.0% [##########] 11.5 MiB (338.0 KiB / s)
[info] Fetched artifacts of 
[info] compiling 12 Scala sources to /Users/williamcheng/Code/openapi-generator7/samples/client/petstore/scala/sttp4/target/scala-2.13/classes ...
https://repo1.maven.org/maven2/org/scala-sbt/compiler-bridge_2.13/1.9.0/compiler-bridge_2.13-1.9.0.pom
  100.0% [##########] 2.7 KiB (15.9 KiB / s)
[info] Non-compiled module 'compiler-bridge_2.13' for Scala 2.13.10. Compiling...
[info]   Compilation completed in 24.446s.
[success] Total time: 75 s (01:15), completed 21 Jun. 2023, 4:39:10 pm
```

**Comment by wing328** [comment]
@flsh86 when you've time, can you please add the new sample folder to `.github/workflows/samples-scala.yaml` so that the CI will test any change to scala-sttp4 petstore client moving forward?

**Comment by wing328** [comment]
More of an FYI. We're moving away from Petstore server to Echo server for client testing. For more information, please refer to https://github.com/OpenAPITools/openapi-generator/wiki/Integration-Tests#echo-server. You may find it useful in adding tests with Echo server to cover cases not yet supported by Petstore server.

File: samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala (hunk 6 above)
Comment (wing328) [reply samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala:23]: will try to fix the formatting issue after merging this PR.

Comment (wing328) [reply samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala:23]: fixed via https://github.com/OpenAPITools/openapi-generator/pull/15931

File: samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala (hunk 5 above)
Comment (wing328) [reply samples/client/petstore/scala/sttp4/src/main/scala/org/openapitools/client/api/PetApi.scala:54]: fixed via https://github.com/OpenAPITools/openapi-generator/pull/15931

**Comment by wing328** [comment]
@fish86 can you please PM me via Slack when you've time?

https://join.slack.com/t/openapi-generator/shared_invite/zt-12jxxd7p2-XUeQM~4pzsU9x~eGLQqX2g

**Pull request merged**